storage/olddbclean/oldDatabaseCleaner.go

```diff
@@ -0,0 +1,231 @@
+package olddbclean
+
+import (
+	"math"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/epochStart/notifier"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/storage/factory/directoryhandler"
+)
+
+var log = logger.GetOrCreate("storage/olddbclean")
+
+// Args holds the arguments needed for creating an oldDatabaseCleaner
+type Args struct {
+	DatabasePath        string
+	StorageListProvider StorageListProviderHandler
+	EpochStartNotifier  EpochStartNotifier
+}
+
+type oldDatabaseCleaner struct {
+	databasePath        string
+	storageListProvider StorageListProviderHandler
+	fileRemover         func(file string) error
+	directoryReader     storage.DirectoryReaderHandler
+	oldestEpochsToKeep  map[uint32]uint32
+
+	sync.RWMutex
+}
+
+// NewOldDatabaseCleaner returns a new instance of oldDatabaseCleaner
+func NewOldDatabaseCleaner(args Args) (*oldDatabaseCleaner, error) {
+	if check.IfNil(args.StorageListProvider) {
+		return nil, storage.ErrNilStorageListProvider
+	}
+	if check.IfNil(args.EpochStartNotifier) {
+		return nil, storage.ErrNilEpochStartNotifier
+	}
+
+	fileRemoverFunc := func(file string) error {
+		return os.RemoveAll(file)
+	}
+	directoryReader := directoryhandler.NewDirectoryReader()
+
+	odc := &oldDatabaseCleaner{
+		databasePath:        args.DatabasePath,
+		storageListProvider: args.StorageListProvider,
+		fileRemover:         fileRemoverFunc,
+		directoryReader:     directoryReader,
+		oldestEpochsToKeep:  make(map[uint32]uint32),
+	}
+
+	odc.registerHandler(args.EpochStartNotifier)
+
+	return odc, nil
+}
+
+// registerHandler will register a new function to the epoch start notifier
+func (odc *oldDatabaseCleaner) registerHandler(handler EpochStartNotifier) {
+	subscribeHandler := notifier.NewHandlerForEpochStart(
+		func(hdr data.HeaderHandler) {
+			err := odc.handleEpochChangePrepare(hdr.GetEpoch())
```

Comment:
confusion: this is not the prepare but action. Please rename. Because of this bad naming I thought you wired to the prepare event which would have been a mistake.

Reply:
Yes. For the first implementation I was registering to EpochStartPrepare but later refactreod, so that's the naming reason. Renamed it